internal/httpapi/server.go

```diff
@@ -90,6 +101,49 @@ func (c *ServerController) HandleRegister(ctx *gin.Context) {
 	ctx.JSON(http.StatusCreated, gin.H{"result": "registered"})
 }
 
+// HandleRegister is a gin HTTP handler that allows servers to register
+// themselves in the repository. This call will also dial back to the port
+// being registered and confirm that the port is accessible.
+func (c *ServerController) HandleRegister(ctx *gin.Context) {
+	serverAddr, err := srvrepo.ParseServerAddress(ctx.Param("server_id"))
+	if err != nil {
+		// 404, since the ID is a URL param
+		ctx.JSON(http.StatusNotFound, gin.H{"result": "invalid server ID"})
+		return
+	}
+
+	// If we have already seen this server, just update it
+	existed, err := c.repository.Has(srvrepo.ServerID(serverAddr.String()))
+	if existed {
+		c.HandleUpdate(ctx)
+		return
+	}
+
+	destinationAddress, _ := net.ResolveUDPAddr("udp", serverAddr.String())
+	connection, err := net.DialUDP("udp", nil, destinationAddress)
+	defer connection.Close()
+	if err != nil {
+		log.Fatal(err)
+		ctx.JSON(http.StatusPreconditionFailed, gin.H{"result": "Repository could not ping you."})
+	}
+
+	var buffer bytes.Buffer
+	buffer.WriteString("ping")
+	for ii := 0; ii < 10; ii++ {
```

Comment:
Drop a comment in here noting we're trying a handful due to the fact that we're dealing with an unreliable transport protocol.

You know, so the thousands of people who will read this won't get confused.